python/src/graphql.rs

```diff
@@ -1,87 +1,590 @@
+
+        match graphql_result.remove("data") {
+            Some(Value::Object(data)) => Ok(data.into_iter().collect()),
+            _ => match graphql_result.remove("errors") {
+                Some(Value::Array(errors)) => Err(PyException::new_err(format!(
+                    "Got errors from GraphQL server:\n{errors:?}"
+                ))),
+                _ => Err(PyException::new_err(format!(
+                    "Error while reading server response for query:\n\t{query}"
+                ))),
+            },
         }
-        (None, None) => server,
-        _ => Err(PyException::new_err(
-            "invalid arguments, you have to set both vectorise and cache accordingly",
-        ))?,
-    };
-
-    tokio::future_into_py(py, async move {
-        vector_server
-            .run_with_port(port)
+    }
+
+    async fn send_graphql_query(
+        &self,
+        query: String,
+        variables: HashMap<String, Value>,
+    ) -> PyResult<HashMap<String, Value>> {
+        let client = Client::new();
+
+        let request_body = json!({
+            "query": query,
+            "variables": variables
+        });
+
+        let response = client
+            .post(&self.url)
+            .json(&request_body)
+            .send()
             .await
-            .map_err(|e| adapt_err_value(&e))
-    })
+            .map_err(|err| adapt_err_value(&err))?;
+
+        response.json().await.map_err(|err| adapt_err_value(&err))
+    }
+
+    fn generic_load_graphs(
+        &self,
+        py: Python,
+        load_function: &str,
+        path: String,
+    ) -> PyResult<HashMap<String, PyObject>> {
+        let query =
+            format!("mutation LoadGraphs($path: String!) {{ {load_function}(path: $path) }}");
+        let variables = [("path".to_owned(), json!(path))];
+
+        let data = self.query_with_json_variables(query.clone(), variables.into())?;
+
+        match data.get(load_function) {
+            Some(Value::Array(loads)) => {
+                let num_graphs = loads.len();
+                println!("Loaded {num_graphs} graph(s)");
+                translate_map_to_python(py, data)
+            }
+            _ => Err(PyException::new_err(format!(
+                "Error while reading server response for query:\n\t{query}\nGot data:\n\t'{data:?}'"
+            ))),
+        }
+    }
+
+    fn is_online(&self) -> bool {
+        match reqwest::blocking::get(&self.url) {
+            Ok(response) => response.status().as_u16() == 200,
+            _ => false,
+        }
+    }
 }
 
-#[pyfunction]
-pub fn from_map_and_directory(
+const WAIT_CHECK_INTERVAL_MILLIS: u64 = 200;
+
+#[pymethods]
+impl PyRaphtoryClient {
+    #[new]
+    fn new(url: String) -> Self {
+        Self { url }
+    }
+
+    /// Wait for the server to be online.
+    ///
+    /// Arguments:
+    ///   * `millis`: the minimum number of milliseconds to wait (default 5000).
+    fn wait_for_online(&self, millis: Option<u64>) -> PyResult<()> {
```

Comment:
Its the maximum time to wait no?

Reply:
it is the minimum number of milliseconds to wait "if the server doesn't respond", because if you set 250 it might wait for longer, but I'll change it to just "the timeout in milliseconds (default 5000)"